src/oss/langchain/multi-agent.mdx

```diff
@@ -1141,143 +1141,226 @@ This is how handoffs via tools work in multi-agent systems. It's similar to the
+    response = model.with_structured_output(RewrittenQuery).invoke([
+        {"role": "system", "content": system_prompt},
+        {"role": "user", "content": state["question"]}
+    ])
+    return {"rewritten_query": response.query}
 
-    function extractText(state: typeof DocumentState.State) {
-      const raw = state.document;
-      return { extractedText: `[Extracted from document]: ${raw}` };
-    }
+def retrieve(state: State) -> dict:
+    """Retrieve documents based on the rewritten query."""
+    docs = retriever.invoke(state["rewritten_query"])
+    return {"documents": [doc.page_content for doc in docs]}
 
-    const summarizeAgent = createAgent({
-      model: "openai:gpt-4o",
-      prompt: "Summarize the following text concisely.",
-      name: "summarizer"
-    });
+def call_agent(state: State) -> dict:
+    """Generate answer using retrieved context."""
+    context = "\n\n".join(state["documents"])
+    prompt = f"Context:\n{context}\n\nQuestion: {state['question']}"
+    response = agent.invoke({"messages": [{"role": "user", "content": prompt}]})
+    return {"answer": response["messages"][-1].content}
 
-    async function summarize(state: typeof DocumentState.State) {
-      const result = await summarizeAgent.invoke({
-        messages: [{ role: "user", content: state.extractedText }]
-      });
-      return { summary: result.messages.at(-1)?.content };
-    }
+workflow = (
+    StateGraph(State)
+    .add_node("rewrite", rewrite_query)
+    .add_node("retrieve", retrieve)
+    .add_node("agent", call_agent)
+    .add_edge(START, "rewrite")
+    .add_edge("rewrite", "retrieve")
+    .add_edge("retrieve", "agent")
+    .add_edge("agent", END)
+    .compile()
+)
 
-    const translateAgent = createAgent({
-      model: "openai:gpt-4o",
-      prompt: "Translate the following text to Spanish.",
-      name: "translator"
-    });
+result = workflow.invoke({"question": "Who won the 2024 WNBA Championship?"})
+print(result["answer"])
+```
+:::
+:::js
+```typescript
+import { StateGraph, Annotation, START, END } from "@langchain/langgraph";
+import { createAgent, tool } from "langchain";
+import { ChatOpenAI, OpenAIEmbeddings } from "@langchain/openai";
+import { MemoryVectorStore } from "langchain/vectorstores/memory";
+import * as z from "zod";
 
-    async function translate(state: typeof DocumentState.State) {
-      const result = await translateAgent.invoke({
-        messages: [{ role: "user", content: state.summary }]
-      });
-      return { translation: result.messages.at(-1)?.content };
-    }
+const State = Annotation.Root({
+  question: Annotation<string>(),
+  rewrittenQuery: Annotation<string>(),
+  documents: Annotation<string[]>(),
+  answer: Annotation<string>(),
+});
+
+// WNBA knowledge base with rosters, game results, and player stats
+const embeddings = new OpenAIEmbeddings();
+const vectorStore = await MemoryVectorStore.fromTexts(
```

Comment:
note: We should be showing `from_documents` generally rather than from_texts (not a blocker)